x-pack/plugins/apm/server/utils/source_map_rt.ts

```diff
@@ -0,0 +1,42 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+import { either } from 'fp-ts/lib/Either';
+import * as t from 'io-ts';
+
+const sourceMapRt = t.intersection([
+  t.type({
+    version: t.number,
+    sources: t.array(t.string),
+    mappings: t.string,
+  }),
+  t.partial({
+    names: t.array(t.string),
+    file: t.string,
+    sourceRoot: t.string,
+    sourcesContent: t.array(t.string),
+  }),
+]);
+
+export type SourceMap = t.TypeOf<typeof sourceMapRt>;
+
+export const toSourceMapRt = new t.Type<SourceMap, Buffer | string, unknown>(
```

Comment:
I think what you need here is something like:

`t.union([ t.string, bufferAsStringRt ]).pipe(sourceMapRt)`

Reply:
or maybe `t.union([ sourceMapRt, bufferAsStringRt.pipe(sourceMapRt) ])`